README.md
[![Build Status](https://secure.travis-ci.org/vesln/seed-forge.png)](http://travis-ci.org/vesln/seed-forge)

# Seed Forge

Seed Forge is a powerful factory builder for [Seed](https://github.com/qualiancy/seed).

Seed is a storage-agnostic graphing database framework that provides a common API for working
with JSON-style documents.

## Synopsis

### Public methods

```js
var define     = require('seed-forge').define;
var list       = require('seed-forge').list;
var build      = require('seed-forge').build;
var factory    = require('seed-forge').factory;
var attributes = require('seed-forge').attributes;
```

### Define a factory

```js
define(User)
  .set('name', 'Name')
  .set('url', function () {
    return 'http://example.com';
  })
  .set('eyes.left', 'blue')
  .set('eyes.right', 'green')
  .set('email', function (n) {
    return 'example' + n + '@example.com'
  })
  .set('age', 33)
  .set('admin', false);
```

### Inheritance

```js
define('Admin', User)
  .extend('User')
  .set('age', 44)
  .set('admin', true)
```

### Create a record from factory

```js
factory('User', fn);
factory('User', { override: 'attributes' }, fn);
```

### Create multiple records

```js
list('User', 10, fn); // Will create 10 users.
```

### Build a record, without saving it

```js
var user = build('User');
var user = build('User', { override: 'attributes' });
```

### Get model attributes from a factory

```js
var attrs = attributes('User');
```

## Requirements

- NPM (http://npmjs.org/)
- Node.js (http://nodejs.org/) >= 0.6.0

## Install

```
$ npm install seed-forge
```

## Tests

```
$ npm install
$ make test
```

## TODO

- Better docs

## License

MIT License

Copyright (C) 2012 Veselin Todorov (hi@vesln.com)

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


History.md

0.0.2 / 2012-10-05
==================

  * Use actual Schema

0.0.1 / 2012-10-05
==================

  * Initial version


test/support/models.js
/*!
 * Seed Forge - factories for Seed.
 *
 * Veselin Todorov <hi@vesln.com>
 * MIT License
 */

/**
 * Seed.
 *
 * @type {Object}
 */

var seed = require('seed');

/**
 * Memory Storage.
 *
 * @type {Object}
 */

var store = new seed.MemoryStore();

/**
 * Example user model.
 *
 * @type {Function}
 */

module.exports.User = seed.Model.extend('User', {
  store: store,
  schema: new seed.Schema({
    name: String,
    url: String,
    eyes: {
      left: String,
      right: String,
    },
    age: Number,
    email: String,
    admin: Boolean,
  })
});


/**
 * Invalid model (without a storage).
 *
 * @type {Function}
 */

module.exports.Invalid = seed.Model.extend('Invalid', {});


test/seed-forge.test.js
/*!
 * Seed Forge - factories for Seed.
 *
 * Veselin Todorov <hi@vesln.com>
 * MIT License
 */

/**
 * With `list` one can create multiple
 * factories at once.
 *
 * @type {Function}
 */
var list = require('../').list;

/**
 * With `define` one can define a factory and
 * rules for it.
 *
 * @type {Function}
 */
var define = require('../').define;

/**
 * `build` builds a valid model object, but not
 * persisted yet.
 *
 * @type {function}
 */
var build = require('../').build;

/**
 * The factory creator.
 *
 * @type {function}
 */
var factory = require('../').factory;

/**
 * Factory attributes reader.
 *
 * @type {function}
 */
var attributes = require('../').attributes;

/**
 * Invalid model.
 *
 * @type {Function}
 */
var Invalid = require('./support/models').Invalid;

/**
 * Example User model.
 *
 * @type {Function}
 */
var User = require('./support/models').User;

define(User)
  .set('name', 'Name')
  .set('url', function () {
    return 'http://example.com';
  })
  .set('eyes.left', 'blue')
  .set('eyes.right', 'green')
  .set('email', function (n) {
    return 'example' + n + '@example.com'
  })
  .set('age', 33)
  .set('admin', false);

define('Admin', User)
  .extend('User')
  .set('admin', true)

define(Invalid);

describe('Seed Forge', function () {
  describe('overriding default attributes', function () {
    it('can be done when building', function () {
      var user = build('User', { age: 44 });
      user.get('age').should.eq(44);
    });

    it('can be done when creating', function (done) {
      factory('User', { age: 44 }, function (user) {
        user.fetch(function () {
          user.get('age').should.eq(44);
          done();
        });
      });
    });
  });

  describe('creating a factory', function() {
    it('can save the factory', function (done) {
      factory('User', function (user) {
        user.fetch(function () {
          user.get('_id').should.be.not.null;
          user.get('eyes').should.eql({ left: 'blue', right: 'green' });
          user.get('name').should.eq('Name');
          user.get('age').should.eql(33);
          user.get('url').should.eql('http://example.com');

          done();
        });
      });
    });

    it('throws an exception if one occures by saving', function() {
      var fn = function () {
        factory('Invalid', function (invalid) {});
      };

      fn.should.throw();
    });
  });

  it('can return valid object, but not persisted', function () {
    var user = build('User');

    user.get('name').should.eq('Name');
    user.get('age').should.eq(33);
  });

  it('can return valid attributes', function () {
    var attrs = attributes('User');

    attrs.name.should.eq('Name');
    attrs.age.should.eq(33);
    attrs.url.should.eq('http://example.com');
    attrs.eyes.should.eql({ left: 'blue', right: 'green' });
  });

  it('can handle sequences', function () {
    var first = attributes('User')
    , second = attributes('User')
    , regexp = /example[0-9]+@example.com/;

  first.email.should.not.eq(second.email);

  first.email.should.match(regexp);
  second.email.should.match(regexp);
  });

  it('supports inheritance', function () {
    var user = build('User');
    var admin = build('Admin');

    user.get('eyes').should.eql(admin.get('eyes'));
    user.get('name').should.eq(admin.get('name'));
    user.get('age').should.eq(admin.get('age'));

    user.get('admin').should.eq(false);
    admin.get('admin').should.eq(true);

    user.get('email').should.not.eq(admin.get('emai'));
  });

  it('can create multiple factories at once', function (done) {
    list('User', 10, function (users) {
      users.length.should.eq(10);

      users.map(function (user) {
        user.get('_id').should.be.ok;
      });

      done();
    });
  });
});


test/integration.test.js
/*!
 * Seed Forge - factories for Seed.
 *
 * Veselin Todorov <hi@vesln.com>
 * MIT License
 */

/**
 * With `list` one can create multiple
 * factories at once.
 *
 * @type {Function}
 */
var list = require('../').list;

/**
 * With `define` one can define a factory and
 * rules for it.
 *
 * @type {Function}
 */
var define = require('../').define;

/**
 * `build` builds a valid model object, but not
 * persisted yet.
 *
 * @type {function}
 */
var build = require('../').build;

/**
 * The factory creator.
 *
 * @type {function}
 */
var factory = require('../').factory;

/**
 * Factory attributes reader.
 *
 * @type {function}
 */
var attributes = require('../').attributes;

/**
 * Example User model.
 *
 * @type {Function}
 */
var User = require('./support/models').User;

define(User)
  .set('name', 'Name')
  .set('url', function () {
    return 'http://example.com';
  })
  .set('eyes.left', 'blue')
  .set('eyes.right', 'green')
  .set('email', function (n) {
    return 'example' + n + '@example.com'
  })
  .set('age', 33)
  .set('admin', false);

define('Admin', User)
  .extend('User')
  .set('admin', true)

describe('Seed Forge', function () {
  describe('overriding default attributes', function () {
    it('can be done when building', function () {
      var user = build('User', { age: 44 });
      user.get('age').should.eq(44);
    });

    it('can be done when creating', function (done) {
      factory('User', { age: 44 }, function (user) {
        user.fetch(function () {
          user.get('age').should.eq(44);
          done();
        });
      });
    });
  });

  it('can create from factories', function (done) {
    factory('User', function (user) {
      user.fetch(function () {
        user.get('_id').should.be.not.null;
        user.get('eyes').should.eql({ left: 'blue', right: 'green' });
        user.get('name').should.eq('Name');
        user.get('age').should.eql(33);
        user.get('url').should.eql('http://example.com');

        done();
      });
    });
  });

  it('can return valid object, but not persisted', function () {
    var user = build('User');

    user.get('name').should.eq('Name');
    user.get('age').should.eq(33);
  });

  it('can return valid attributes', function () {
    var attrs = attributes('User');

    attrs.name.should.eq('Name');
    attrs.age.should.eq(33);
    attrs.url.should.eq('http://example.com');
    attrs.eyes.should.eql({ left: 'blue', right: 'green' });
  });

  it('can handle sequences', function () {
    var first = attributes('User')
      , second = attributes('User')
      , regexp = /example[0-9]+@example.com/;

    first.email.should.not.eq(second.email);

    first.email.should.match(regexp);
    second.email.should.match(regexp);
  });

  it('supports inheritance', function () {
    var user = build('User');
    var admin = build('Admin');

    user.get('eyes').should.eql(admin.get('eyes'));
    user.get('name').should.eq(admin.get('name'));
    user.get('age').should.eq(admin.get('age'));

    user.get('admin').should.eq(false);
    admin.get('admin').should.eq(true);

    user.get('email').should.not.eq(admin.get('emai'));
  });

  it('can create multiple factories at once', function (done) {
    list('User', 10, function (users) {
      users.length.should.eq(10);

      users.map(function (user) {
        user.get('_id').should.be.ok;
      });

      done();
    });
  });
});

xit('raises error on create');


lib/seed-forge.js
/*!
 * Seed Forge - factories for Seed.
 *
 * Veselin Todorov <hi@vesln.com>
 * MIT License
 */

/*!
 * Internal dependencies.
 */

var Factory = require('./seed-forge/factory');
var manager = require('./seed-forge/manager');

/**
 * Create and save a factory, found by type.
 *
 * @param {String} Factory name
 * @param {Object} Custom attributes [optional]
 * @param {Function} Callback
 * @api public
 */

function factory (name, attrs, fn) {
  if (arguments.length === 2) {
    fn = attrs;
    attrs = null;
  }

  manager.get(name).create(attrs, fn);
};

/**
 * Create an factory, but don't save it.
 *
 * @param {String} Factory name
 * @param {Object} Custom attributes [optional]
 * @api public
 */

function build (name, attrs) {
  return manager.get(name).build(attrs);
};

/**
 * Return valid attributes for a model.
 *
 * @param {String} name
 * @returns {Object}
 * @api public
 */

function attributes (name) {
  return manager.get(name).attrs();
};

/**
 * Define a factory.
 *
 * @param {String} Type [optional]
 * @param {Function} Seed model
 * @api public
 */

function define (type, Model) {
  if (arguments.length === 1) {
    Model = type;
    type = Model.prototype.__type;
  }

  var factory = new Factory(Model);
  manager.set(type, factory)

  return factory;
};


/**
 * Create multiple records at once.
 *
 * @param {String} Factory name
 * @param {Number} The number of record to be created
 * @param {Function} Callback
 * @api public
 */

function list (type, num, fn) {
  var records = [];

  var created = function (record) {
    records.push(record);
    if (records.length === num) fn(records);
  };

  for (var i = 0; i < num; i++) {
    factory(type, created);
  }
};

/*!
 * Module version
 */

module.exports.version = require('../package.json').version;

/*!
 * Expose `define`.
 */

module.exports.define = define;

/**
 * Expose `build`.
 */

module.exports.build = build;

/**
 * Expose `factory`.
 */

module.exports.factory = factory;

/*!
 * An alias for factory.
 */

module.exports.create = factory;

/*!
 * Expose `attributes`.
 */

module.exports.attributes = attributes;

/*!
 * Expose `list`
 */

module.exports.list = list;


lib/seed-forge/factory.js
/*!
 * Seed Forge - factories for Seed.
 *
 * Veselin Todorov <hi@vesln.com>
 * MIT License
 */

/*!
 * Module dependencies.
 */

var merge = require('super')
  , clone = merge
  , filtr = require('filtr');

/*!
 * Internal dependencies.
 */

var manager = require('./manager')
  , raise = require('./utils').raise;

/**
 * The core factory class.
 *
 * @param {Function} Seed model
 * @api public
 */

function Factory (Model) {
  this.Model = Model;
  this.attributes = {};
  this.i = 0;
};

/**
 * Set a value for an attribute.
 *
 * @param {String} name
 * @param {Mixed} value
 * @returns `this`
 * @api public
 */

Factory.prototype.set = function (name, value) {
  this.attributes[name] = value;
  return this;
};

/**
 * Build and save a factory.
 *
 * @param {Object} custom attributes [optional]
 * @param {Function} callback
 * @api public
 */

Factory.prototype.create = function (attrs, fn) {
  var obj = this.build(attrs);

  obj.save(function (err) {
    raise(err);
    fn(obj);
  });
};

/**
 * Build a factory.
 *
 * @param {Object} custom attributes [optional]
 * @returns {Object}
 * @api public
 */

Factory.prototype.build = function (attrs) {
  attrs = merge(
    this.attrs(),
    attrs || {}
  );

  return new this.Model(attrs);
};

/**
 * Return own and parent attributes.
 *
 * @returns {Object} attributes
 * @api public
 */

Factory.prototype.attrs = function () {
  var ret = {}
    , attributes = this.attributes;

  if (this.hasParent()) {
    attributes = merge(this.parent.attributes, attributes);
  }

  Object.keys(attributes).forEach(function (attribute) {
    filtr.setPathValue(
      attribute,
      this.value(attribute, attributes),
      ret
    );
  }.bind(this));

  return ret;
};

/**
 * Extend the current factory.
 *
 * @param {String} Name of the parent factory
 * @returns `this`
 * @api public
 */
Factory.prototype.extend = function (name) {
  this.parent = manager.get(name);
  return this;
};

/**
 * Compute a value for an attribute.
 *
 * @param {String} attribute
 * @param {Object} definition
 * @returns {Mixed}
 * @api private
 */
Factory.prototype.value = function (attribute, definition) {
  if (typeof definition[attribute] !== 'function') {
    return definition[attribute];
  }

  return (definition[attribute].length === 0)
    ? definition[attribute]()
    : definition[attribute](this.seq())
  ;
};

/**
 * Check if the factory has a parent factory.
 *
 * @returns {Boolean}
 * @api private
 */

Factory.prototype.hasParent = function () {
  return !!this.parent;
};

/**
 * Return an integer, unique for the current factory.
 *
 * @returns {Number}
 * @api private
 */

Factory.prototype.seq = function () {
  return (this.hasParent())
    ? this.parent.seq()
    : ++this.i
  ;
};


/*!
 * Expose `Factory`
 */

module.exports = Factory;


lib/seed-forge/manager.js
/*!
 * Seed Forge - factories for Seed.
 *
 * Veselin Todorov <hi@vesln.com>
 * MIT License
 */

/**
 * Factory manager. Store the factories.
 *
 * @api public
 */

function Manager () {
  this.factories = {};
};

/**
 * Set a factory.
 *
 * @param {String} Name
 * @param {Object} Factory
 * @returns `this`
 * @api public
 */

Manager.prototype.set = function (name, factory) {
  return this.factories[name] = factory;
};

/**
 * Get a factory. It throws an exception
 * if the factory is not defined.
 *
 * @param {String} Name
 * @returns {Object}
 * @api public
 */

Manager.prototype.get = function (name) {
  if (!this.factories[name]) {
    throw new Error('The factory "' + name + '" is not defined.');
  }

  return this.factories[name];
};

/*!
 * Expose `Manager`.
 */

module.exports = new Manager;


examples/index.js
/*!
 * Seed Forge - factories for Seed.
 *
 * Veselin Todorov <hi@vesln.com>
 * MIT License
 */

/**
 * Seed.
 *
 * @type {Object}
 */
var seed = require('seed');

/**
 * Memory Storage.
 *
 * @type {Object}
 */

var store = new seed.MemoryStore();

/*!
 * An example callback function used in the examples.
 */
var fn = function () {};

/*!
 * Seed-factory methods.
 */
var define     = require('..').define;
var list       = require('..').list;
var build      = require('..').build;
var factory    = require('..').factory;
var attributes = require('..').attributes;

/**
 * Example User model. This should be your
 * model.
 *
 * @type {Function}
 */
var User = seed.Model.extend('User', {
  store: store,
  schema: new seed.Schema({
    name: String,
    url: String,
    eyes: {
      left: String,
      right: String,
    },
    age: Number,
    email: String,
    admin: Boolean,
  })
});

/*!
 * Define `User` factory.
 *
 * Sequences, static values and dynamic values.
 */

define(User)
  .set('name', 'Name')
  .set('url', function () {
    return 'http://example.com';
  })
  .set('eyes.left', 'blue')
  .set('eyes.right', 'green')
  .set('email', function (n) {
    return 'example' + n + '@example.com'
  })
  .set('age', 33)
  .set('admin', false);

/*!
 * Inheritance.
 */

define('Admin', User)
  .extend('User')
  .set('age', 44)
  .set('admin', true)


/*!
 * Create a factory.
 */

factory('User', fn);
factory('User', { override: 'attributes' }, fn);

/*!
 * Save multiple records at once.
 */

list('User', 10, fn);

/*!
 * Build a factory, without saving it.
 */

var user = build('User');
var user = build('User', { override: 'attributes' });

/*!
 * Get factory attributes.
 */

var attrs = attributes('User');
